### DISCUSSION TITLE
Addon/Script Project: Query Node. Bringing queries to Freeplane.

### DISCUSSION THREAD

**Discussion by euu2021** [freeplane/freeplane#548] [Show and tell]
scriptDescription: allows creating criteria to have nodes be automatically cloned into another node 

Some notetaking programs have a popular feature that is usually called "queries". The idea is: the user defines criteria for the program to transclude itens automatically on a given spot. In FP terms: show the search results (nodes) inside a node that the user chooses.
It's similar to simply filtering, but the user can "embbed" the results into a given place, so it conveniently appears in the proper context.

[This video](https://www.youtube.com/watch?v=HoccqyiHvPw) shows how it works in Roam Research.

Now, beyond this "visualizing a search" aspect, there is also an aspect of grabbing the resulting nodes and organizing them in ways that are different from the original place. For example: grabbing all TODO nodes that are around the map and put them on a Kanban board. For my concrete use case, I want to put all the nodes of a given style into a unified Kanban board AND rank the nodes inside the Kanban board, in order of priority:

My usecase: I organize study subjects inside a tree that follows the traditional textbook structure. Inside each subject, there are some categories of nodes that I want to rank in a unified node. For example, questions that I have and need to find answers for (there are so many that I need to actually rank them, so I can prioritize correctly).

Doing this manipulation of the results of the query is also possible on other programs, but it's usually clumsy. But, with this script below, it's easy to do that in Freeplane. Example:

![javaw_GtUBPWMiI1](https://user-images.githubusercontent.com/77707706/183744511-0b25d0f3-abf0-458e-8ef0-9d561d041428.gif)

The script is:

```
//the variables below are the ones that should be changed as needed

def branchOrigin = ID_1318286144 //ID of the parent node of the branch to be checked
def targetNode = ID_1250375737 //ID of the target node (query node)
def styleTargeted = "name of the style" //style to be met

//first part: cloning the maching nodes into the query node

branchOrigin.findAll().each{ 
	def nodeInside = it
	if (nodeInside.hasStyle(styleTargeted)) { 
		def clonesList1 = [ ]
		nodeInside.getNodesSharingContentAndSubtree().each{
				clonesList1.add(it)
		}
		def targetNodeList = [ ]
		targetNode.findAll().each{
				targetNodeList.add(it)
		}
		if (
			targetNodeList.any { clonesList1.contains(it) }
		) {
		
		}
		
		else {
			targetNode.appendAsCloneWithSubtree(nodeInside)
		}
	}	

	else {
		
	}
}

//second part: excluding from the querie node the nodes excluded from the origin branch

def clonesList2 = [ ]

branchOrigin.findAll().each{ 	
	if (it.hasStyle(styleTargeted)) { 		
		it.getNodesSharingContentAndSubtree().each{
				clonesList2.add(it)
		}
	}
	else {
		
	}
}

targetNode.findAll().each{			
	if (!it.hasStyle(styleTargeted) | clonesList2.contains(it)){
		
	}
	else {
		it.delete()
	}
}
```

So, the 3 variables to be changed are right at the top: the parent node of the branch to be considered; the node that will be the Query Node; the search criteria (in the script below, it's only the Style).

Deleting the original node deletes the node in the Query Node. The check is made every time the script is applied.

I did not find other scripts that do that. If you know any, please show me, so I can learn from it.

---
Wishlist for future improvements:

- [ ] Execute automatically on the background
- [ ] Easy way to choose the origin-branch and the target node (node query)
- [ ] Query builder: something that is easy like the filter composer in FP
- [ ] make tests to see how it works with multiple origin-branches/targets
- [ ] radio buttons to choose one mode of: use clones with branches; use clones without branches; use copy with link
- [ ] options to choose: if the node is deleted on the origin, what should happen to the clone? Also gets deleted; Change the formatting/style, in order to make it clear that the original was deleted; nothing

**Comment by EdoFro** [comment]
@eu2021,
I have something like that.
I use my maps for proyects and I add tasks nodes all over it (GTD like).
I also have "bucket" nodes and a script that collect the tasks in the map and copies clones of them (without children) into the bucket. this way I have a single place were I can see al the pending tasks of the map.

this works also with other styles also.For example, I use it once to collect all nodes that have scripts in the map.

It has also buttons to easy navigate from the selected task in the bucket to the actual node in the map. etc.

**Reply by euu2021** [reply]
Please, if possible, share this script with us. It sounds to be interesting.

**Comment by EdoFro** [comment]
I use this in the [Tutorial-o-Matic project.mm](https://github.com/EdoFro/Freeplane_Tutorial_AddOn/blob/main/Tutorial-o-Matic%20project.mm) map.

Later will I add a simplier demo and the scripts
 bye, 
edo

**Reply by euu2021** [reply]
Please, do. I had to write my script from scratch, so I had lots of fun designing the logic. Now, I want to see how a more experienced user would solve the same problem.

 Tutorial-o-Matic is looking good. It has a lot of potential.

**Comment by macmarrum** [comment]
A good idea, both of you.

@eu2021 I think it's the right direction to have the logic executed automatically (in the background). A MapChangeListener would fit the requirements when a node is moved or deleted; a NodeChangeListener for when the target style is assigned to a (new) node.

**Comment by Illegit** [comment]
@eu2021: 
1) does make it sense to use as a branch-origin node the current node? (if so, how to change the script?)
2) what about if the target node is on another map?
3) if instead of a clone I apply a copy of just the node and the link to the target node?
@eu2021 @macmarrum @EdoFro 
I'm eager to read the evolution of this thread!

**Reply by euu2021** [reply]
> 1. does make it sense to use as a branch-origin node the current node? (if so, how to change the script?)

What you want to achieve?

If the current node is used as the branch-origin node, then the user needs to select that node every time he wants to "refresh" the nodes. If applied on other node, then the clones would be deleted and replaced by the ones of this new node.

To achieve that, simply use `node` as the variable, like this:

`def branchOrigin = node`

I guess what you want is to easily set the origin? I think the easiest way for now is choosing an icon or attribute, and making this script consider as the origin the node that has this icon.

I have as a goal creating an easy way to set the origin, target and condition. I will include this in the wishlist in the first post.

> 2. what about if the target node is on another map?

I never looked at crossmapping, but I'm not sure if it's possible to use clones on different maps. I think a copy with link would work, though.

> 3. if instead of a clone I apply a copy of just the node and the link to the target node?

Then it would be necessary to find a common point between the original and the copy (in the script above, I do that by checking the clones list of each node). I think a way would be comparing the URI of the original nodes, with the URIs pointed by the copies.

To be clear: this is necessary because Freeplane needs to know which nodes in the origin branch were already copied and don't need to be copied again.

If you eventually develop that, please post here, so the users can have this option too.

---
I think that, ideally, this script should develop to a point where it would have a GUI where the user could pick:

- origin branch
- target
- condition to be met

And also radio buttons to choose one mode of:

- use clones with branches
- use clones without branches
- use copy with link

Also, other options like: if the node is deleted on the origin, what should happen to the clone?

- Also gets deleted
- Change the formatting/style, in order to make it clear that the original was deleted
- nothing

**Reply by Illegit** [reply]
@eu2021 Thank you.
About 1. you are right, that is what I was asking for.
About 2. and 3.: I cannot promise but I start thinking about it.
